pcdsutils/ophyd_helpers.py

```diff
@@ -0,0 +1,242 @@
+    filter_by : callable, optional
+        Filter signals, with signature ``callable(ophyd.Device.ComponentWalk)``
+    """
+    def default_filter_by(*_) -> bool:
+        return True
+
+    filter_func = filter_by or default_filter_by
+
+    def _get_signals():
+        return [
+            walk.item
+            for walk in device.walk_signals(include_lazy=include_lazy)
+            if filter_func(walk)
+        ]
+
+    if not include_lazy:
+        return _get_signals()
+
+    with no_device_lazy_load():
+        return _get_signals()
+
+
+@contextlib.contextmanager
+def subscription_context_device(
+    device: ophyd.Device,
+    callback: SubscribeCallback,
+    event_type: Optional[str] = None,
+    run: bool = True,
+    *,
+    include_lazy: bool = False,
+    filter_by: Optional[FilterBy] = None
+):
+    """
+    [Context manager] Subscribe to ``event_type`` from signals in ``device``.
+
+    Unsubscribes all signals before exiting
+
+    Parameters
+    ----------
+    device : ophyd.Device
+        ophyd Device to monitor
+    callback : callable
+        Callback to run, with same signature as that of
+        :meth:`ophyd.OphydObj.subscribe`
+    event_type : str, optional
+        The event type to subscribe to
+    run : bool, optional
+        Run the previously cached subscription immediately
+    include_lazy : bool, optional
+        Include lazy signals as well
+    filter_by : callable, optional
+        Filter signals, with signature ``callable(ophyd.Device.ComponentWalk)``
+    """
+    signals = get_all_signals_from_device(
+        device, include_lazy=include_lazy, filter_by=filter_by
+    )
+    with subscription_context(
+        *signals, callback=callback, event_type=event_type, run=run
+    ) as obj_to_cid:
+        yield obj_to_cid
+
+
+@contextlib.contextmanager
+def _acquire(signal: ophyd.Signal):
+    """
+    [Context manager] Subscribe to signal, acquire data until the block exits.
+
+    Parameters
+    ----------
+    signal : ophyd.Signal
+        Ophyd object to monitor.
+
+    Returns
+    -------
+    data : List[PrimitiveType]
+        The data acquired.  Guaranteed to have at least one item.
+    """
+    signal.wait_for_connection()
+    data = []
+
+    start_value = signal.get()
+
+    def acquire(value, **_):
+        data.append(value)
+
+    with subscription_context(signal, callback=acquire):
+        yield data
+
+    if not data:
+        data.extend([start_value, signal.get()])
+
+
+def acquire_blocking(
+    signal: ophyd.Signal,
+    duration: Number
+) -> List[PrimitiveType]:
+    """
+    Subscribe to signal, acquire data for ``duration`` seconds.
+
+    Parameters
+    ----------
+    signal : ophyd.Signal
+        Ophyd object to monitor.
+
+    duration : number
+        Seconds to acquire for.
+
+    Returns
+    -------
+    data : List[PrimitiveType]
+        The data acquired.  Guaranteed to have at least one item.
+    """
+    with _acquire(signal) as data:
```

Comment:
I didn't know you could accumulate generator values into lists like this

Reply:
Maybe we're saying the same thing, but to clarify: the context manager returns a mutable value `data` that's a list - which is appended to and returned by `acquire`.

Reply:
Ah, I misunderstood- thank you for the clarification.
So, the way this works is that we `yield` a list which continues to be mutated until we drop out of this `with` block.